### PR TITLE
[IMP] calendar: schedule reminders using triggers

### PR DESCRIPTION
**Pull request by Julien00859**
The calendar reminders are emails or web notifications sent moments
before an event start. The mechanism to send web notifications is
controlled by the browser, it fetches all the future notifications and
uses `setTimeout()` to delay them. The mechanism to send emails was
controlled by a cron, every 30 minutes the cron would look for the
coming events and send the mail reminder.

The cron was running every 30 minutes even if there was no event. The
best precision was 30 minutes.

The new mechanism use cron triggers, every time one add email reminders,
triggers are created to call the cron at the precise moment the alarm is
set. The cron no longer send emails for events in the coming 30 minutes
as those will be triggered too.

---

**[REF] calendar: remove cron activation code**

There is code that automatically enable or disable the cron whether it
exists alarms. As alarms are automatically added, the cron is actually
always enabled.

Task: 2416741

### CODE REVIEW

File: addons/calendar/models/calendar_event.py [hunk 1]
```diff
@@ -741,12 +741,12 @@ def create(self, vals_list):
...
-            for event in events:
-                if len(event.alarm_ids) > 0:
-                    self.env['calendar.alarm_manager']._notify_next_alarm(event.partner_ids.ids)
+        cron = self.env.ref('calendar.ir_cron_scheduler_alarm')
+        for event in events:
+            for alarm in event.alarm_ids:
+                if alarm.alarm_type == 'email':
+                    cron._trigger(at=event.start-timedelta(minutes=alarm.duration_minutes))
+
         return events
```
Comment (LucasLefevre) [inline addons/calendar/models/calendar_event.py:750]: What about moving those nested (and duplicated) stuffs in a method ?
```py
events._setup_alarms()

def _setup_alarms(self):
    cron = self.env.ref('calendar.ir_cron_scheduler_alarm')
    if not cron:
        return
    for event in self:
        for alarm in event.alarm_ids:
            if alarm.alarm_type == 'email':
                cron._trigger(at=event.start-timedelta(minutes=alarm.duration_minutes))
```

File: addons/calendar/models/calendar_event.py [hunk 2]
```diff
@@ -328,7 +328,7 @@ def _compute_dates(self):
 
     @api.depends('stop', 'start')
     def _compute_duration(self):
-        for event in self.with_context(dont_notify=True):
```
Comment (LucasLefevre) [inline addons/calendar/models/calendar_event.py:331]: :+1:

**Review by LucasLefevre — COMMENTED** [review]
Woop woop

File: addons/calendar/models/calendar_alarm_manager.py [hunk 3]
```diff
@@ -115,55 +116,53 @@ def do_check_alarm_for_one_date(self, one_date, event, event_maxdelta, in_the_ne
...
+                   "alarm"."alarm_type" = 'email'
+               AND "event"."start" - CAST("alarm"."duration" || ' ' || "alarm"."interval" AS Interval) >= %s
+               AND "event"."start" - CAST("alarm"."duration" || ' ' || "alarm"."interval" AS Interval) < now() at time zone 'utc'
+             )''', [self.env.context['lastcall']])
+
+        self.env["calendar.attendee"].search([
+            ("event_id", "in", self.env.cr.fetchall()),
+            ("state", "!=", "declined"),
+            *[("partner_id", "in", partners.ids)
+              for _ in (1,) if partners is not None]
```
Comment (rco-odoo) [inline addons/calendar/models/calendar_alarm_manager.py:160]: ahem :-/

**Review by rco-odoo — APPROVED** [review]
@robodoo delegate+

File: addons/calendar/models/calendar_alarm_manager.py [hunk 4]
```diff
@@ -116,55 +116,54 @@ def do_check_alarm_for_one_date(self, one_date, event, event_maxdelta, in_the_ne
...
+              FROM "calendar_event" AS "event"
+              JOIN "calendar_alarm_calendar_event_rel" AS "event_alarm_rel"
+                ON "event"."id" = "event_alarm_rel"."calendar_event_id"
+              JOIN "calendar_alarm" AS "alarm"
+                ON "event_alarm_rel"."calendar_alarm_id" = "alarm"."id"
+             WHERE (
+                   "alarm"."alarm_type" = 'email'
+               AND "event"."start" - CAST("alarm"."duration" || ' ' || "alarm"."interval" AS Interval) >= %s
+               AND "event"."start" - CAST("alarm"."duration" || ' ' || "alarm"."interval" AS Interval) < now() at time zone 'utc'
+             )''', [self.env.context['lastcall']])
```
Comment (LucasLefevre) [inline addons/calendar/models/calendar_alarm_manager.py:152]: `WHERE event.active = TRUE`

**Review by LucasLefevre — APPROVED** [review]

**Comment by Julien00859** [comment]
Confirmé avec la PO

**Comment by Julien00859** [comment]
@robodoo r+

**Comment by rco-odoo** [comment]
@robodoo rebase-ff

**Pull request closed without merging**